src/app/GitUI/UserControls/MultiSelectTreeView.cs

```diff
@@ -0,0 +1,188 @@
+                value?.EnsureVisible();
+            }
+            finally
+            {
+                _settingFocusedNode = false;
+            }
+        }
+    }
+
+    /// <summary>
+    /// <para>For practical purposes: The last <see cref="TreeNode"/> added to selection.</para>
+    /// <para>Actually: Focused item if selected, otherwise last item in <see cref="SelectedNodes"/> list.</para>
+    /// </summary>
+    [Browsable(false)]
+    [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
+    public TreeNode? LastSelectedNode
+        => _selectedNodes.Contains(base.SelectedNode) ? base.SelectedNode : _selectedNodes.LastOrDefault();
+
+    [Browsable(false)]
+    [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
+    public new TreeNode? SelectedNode
+    {
+        set
+        {
+            _selectedNodes = value is null ? [] : [value];
+            if (value is not null)
+            {
+                FocusedNode = value;
+            }
+
+            Invalidate();
+        }
+    }
+
+    [Browsable(false)]
+    [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
+    public HashSet<TreeNode> SelectedNodes
+    {
+        get => _selectedNodes;
+        set
+        {
+            _selectedNodes = value;
+            if (_selectedNodes.Count > 0)
+            {
+                FocusedNode = _selectedNodes.Last();
+            }
+
+            Invalidate();
+        }
+    }
+
+    public void AddToSelection(TreeNode node)
+    {
+        _selectedNodes.Add(node);
+        FocusedNode = node;
+        Invalidate();
+    }
+
+    protected override void OnEnter(EventArgs e)
+    {
+        base.OnEnter(e);
+        Invalidate();
+    }
+
+    protected override void OnLeave(EventArgs e)
+    {
+        base.OnLeave(e);
+        Invalidate();
+    }
+
+    protected override void OnMouseDown(MouseEventArgs e)
+    {
+        if (e.Button == MouseButtons.Left)
+        {
+            TreeViewHitTestInfo hitTestInfo = HitTest(e.Location);
+            if (hitTestInfo.Node == base.SelectedNode)
+            {
+                // Enforce selection event
+                FocusedNode = null;
+            }
+        }
+
+        base.OnMouseDown(e);
+    }
+
+    private void BeforeSelectHandler(object? sender, TreeViewCancelEventArgs e)
+    {
+        if (!_settingFocusedNode)
+        {
+            Keys modifierKeys = ModifierKeys;
+            if ((modifierKeys | Keys.Control | Keys.Shift) != (Keys.Control | Keys.Shift))
+            {
+                e.Cancel = true;
+                return;
+            }
+
+            UpdateSelection(replace: !modifierKeys.HasFlag(Keys.Control), addRange: modifierKeys.HasFlag(Keys.Shift));
+            Invalidate();
+        }
+
+        FocusedNodeChanged?.Invoke(sender, e);
+
+        return;
+
+        void UpdateSelection(bool replace, bool addRange)
+        {
+            TreeNode newFocusedNode = e.Node;
```

Comment:
It starts from the focused node - like in ListView and Explorer.

Reply:
The focused node is in this implementation updated every shift-click, not like that in explorer or listview.

Reply:
> The focused node is in this implementation updated every shift-click, not like that in explorer or listview.

I see. It will be not easy to change this behavior I guess. I consider multiple shift-clicking a minor edge case.

Reply:
You could do shift-click instead of ctrl-click to extend.
At least add a comment in the code about this unusual behavior, try describe it as a feature...